src/Interpreters/InterpreterAlterQuery.cpp

```diff
@@ -143,6 +143,8 @@ BlockIO InterpreterAlterQuery::executeToTable(const ASTAlterQuery & alter)
 
     if (!mutation_commands.empty())
     {
+        if (alter.settings_ast)
+            mutation_commands.settings_ast = alter.settings_ast->clone();
```

Comment:
`settings_ast` is a member of type `ASTPtr` (= a `shared_ptr<IAST>`) stored in `ASTQueryWithOutput` which is a superclass of `alter`.  We could use the fact that `IAST` inherits from `std::enable_shared_from_this<IAST>`. This means we could dereference `settings_ast` and call `shared_from_this()` which is neatly provided as `ptr()` in class `IAST` already. As a result, we'd only increment a refcount instead of cloning the whole thing.

Perhaps we could do that even unconditionally, i.e. w/o l. 146?

But what I am really thinking to myself is if it would not be better to do the evaluation of the (per-query/volative) setting on this level and instead simply pass a bool down into / store a bool in `MutationCommand`? Having multiple fragments of the same AST in memory seems rather fragile in the long run (even with refcounting).

Reply:
How about just unconditionally doing `mutation_commands.settings_ast = alter.settings_ast;` ?

Since this PR changes the mutation file format, I'm leaning towards a more generic approach than just passing down a bool. It's not hard to imagine other similar cases in the future, and it's better to change the file format less often, IMHO.

Reply:
True, the PR changes the file format. We need to be careful here. I talked to my colleagues and there is an expectation of backwards compatibility. Basically, new versions (which include this PR) should be able to deal with old on-disk formats. The code that reads the persisted data is implemented in the other constructor. Right now, it has format version 1 hardcoded.

@tavplubix Hi Alexander ... Nikolai suggested you may also want to have a look at this PR. We asked ourselves if it is okay to store an ASTPtr in MergeTreeMutationEntry or if there should not be a "settings" struct instead?

@stigsb I think if the new parameter to enable updating materialized columns can *only* be set per-session or globally and not per-query, we wouldn't need to persist it at all, and can save all compatibility discussion 😃

Reply:
It's more or less ok to store `ASTPtr`. It's better to store `SettingsChanges` instead, but anyway the simplest way to (de)serialize it is to use `ASTSettings`, we have similar code in `DDLLogEntry::toString`/`DDLLogEntry::parse`.

However, I'm not sure that we should simply copy settings from `ASTAlterQuery` to mutation entry at all. Because in general ALTER query may have settings that are completely irrelevant for mutation entry:
```
ALTER TABLE t
    UPDATE c = c + 1 WHERE 1
SETTINGS log_queries = 0, mutations_sync = 2, replication_wait_for_inactive_replica_timeout = 300
```
These settings are not needed in `MergeTreeMutationEntry`, but current implementation writes them anyway. Maybe it's not a big issue to have redundant settings in  `MergeTreeMutationEntry`, because these settings will not affect mutation commands execution. But we should notice that this change actually allows to pass an arbitrary setting to a mutation and some settings (like `decimal_check_overflow`, `result_overflow_mode`, `max_memory_usage`, etc) may affect it. 

Also notice that this PR allows to pass settings for simple `MergeTree` only (for `ReplicatedMergeTree` see `ReplicatedMergeTreeMutationEntry`).

Reply:
Okay, so it sounds like the best approach is to use `SettingsChanges` instead of `ASTPtr`, and to populate with specific settings that differ from their default value.  WDYT @tavplubix ?

I will implement for `ReplicatedMergeTree` as well.

Reply:
Maybe it's ok to save changed settings only and ignore all settings except explicitly listed ones (so for now we will save into mutation entry only `update_allow_materialized_columns` and only if it's changed)

Reply:
I have changed the PR to using `SettingsChanges` instead of `ASTPtr` and added support for `ReplicatedMergeTree`.